{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update your sequence alignment code to run as an array job in Slurm. You program should partition the database based on the number of nodes you are using. Then run each of the databases against your query. When all the jobs are completed, coallate the results into a single file.\n",
    "\n",
    "You should use a separate helper script to take care of the partitioning and submitting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have seen firsthand how shared computing resources have both their good and bad points. Fortunately, everyone can now have their own private computing clusters (as long as you're willing to pay for it).\n",
    "\n",
    "Follow this codelab frrom Google Clould Platform to deploy your own HPC Cluster with Slurm: https://codelabs.developers.google.com/codelabs/hpc-slurm-on-gcp/#0\n",
    "\n",
    "Once you have run the sample, get your sequence alignment code to run on it. Provide the sbatch script used to conduct a database search using 4 nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify a gene you are interested in studying for your gene presentation. Ideally, your gene should have a three-dimensional model so that you can do a full analysis of it. However, if it does not, there are options we will discuss in the next class."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Provide a link to the gene in NCBI:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Provide a link to the structure in PDB:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing Proteins \n",
    "\n",
    "The goal of this question is to familiarize yourself with protein structures and to visualize 3D models of proteins using PyMOL.\n",
    "\n",
    "* Download and install PyMOL (https://www.pymol.org/) on your computer. You can download the installer or use Anaconda: `conda install -c schrodinger pymol`. You will not need a license to run it.\n",
    "* Consult the PyMOL wiki (https://pymolwiki.org) for answers to all your questions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download a protein from the PDB within Pymol interactive console using the following command and answer the following questions:\n",
    "```\n",
    "PyMOL>fetch 1mbn\n",
    "```\n",
    "- What is this protein? \n",
    "- What organism does it come from? \n",
    "- Provide the URL to the protein from the PDB website?\n",
    "- Take a screenshot and include it below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you load a structure, the default visualization in PyMOL is a `cartoon` represenation. You can highlight the secondary strucure elements by coloring by secondary structure elements by selecting `C -> by ss -> Helix Sheet Loop`. Answer the following questions:\n",
    "\n",
    "- How many alpha-helices does the protein have? \n",
    "- How many beta sheets does it have?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While this visualization makes for great images for magazine covers, it does not provide the detail we may need to analyze it. In particular, the `licorice sticks` view allows us to see an all-atom reprentation that will help us view the side chains. Change the view in the GUI by selecting `H -> everything` and then `S -> sticks`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer, you can also control the depiction from the command line.\n",
    "```\n",
    "PyMOL> hide everything, 1mbn\n",
    "PyMOL> show sticks, 1mbn\n",
    "```\n",
    "Take a screenshot and include it below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Insert image here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing Proteins\n",
    "\n",
    "Download a different protein from the PDB using the following commands.\n",
    "```\n",
    "PyMOL> fetch 1ase\n",
    "PyMOL> hide everything \n",
    "PyMOL> show cartoon\n",
    "```\n",
    "\n",
    "Answer the following questions.\n",
    "\n",
    "- What is this protein?\n",
    "- How many alpha-helices does the protein have?\n",
    "- How many beta sheets does it have?\n",
    "- What are the ligands (small molecules not part of the protein) that is bound to it?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This protein was part of a functional study that included mutating residues across multiple protein models. Doing this allowed the researchers to identify precisely the amino acids that were functionally active. \n",
    "\n",
    "Load another version of the protein where a mutation was introduced in residue 226.\n",
    "\n",
    "```\n",
    "PyMOL> fetch 1asf\n",
    "PyMOL> align 1ase, 1asf\n",
    "PyMOL> select mutant, resi 226\n",
    "PyMol> show sticks, mutant\n",
    "```\n",
    "\n",
    "- What is the mutation (i.e. what is the change) that was made in the protein? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PyRosetta is a powerful molecular modeling toolkit that we will be using later on in the course. Visualization is a key component to interpreting and analyzing molecular models so they provide a very detailed tutorial as part of their instruction manual. \n",
    "\n",
    "Complete the following PyMOL Tutorial from the PyRosetta group: https://graylab.jhu.edu/pyrosetta/downloads/documentation/pyrosetta4_online_format/PyRosetta4_Workshop1_PyMOL.pdf \n",
    "\n",
    "There are 8 questions in the tutorial, please answer them below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the most common tasks for performing structure analaysis is to performing structure alignments to identify the similarity between protein. You will write a program to perform both global and local structure alignments.  \n",
    "\n",
    "There are many approaches to performing structure alignements, but you will consult [this approach to structure alignment and RMSD](http://boscoh.com/protein/rmsd-root-mean-square-deviation.html) for the technical details on performing the alignments.\n",
    "\n",
    "Despite the complexitity of the underlying approach, you will find it boils down to a couple of matrix command in `numpy`.  As is often the case in bioinformatics, you will be building on what others have contributed to the field.  Understanding other people's code is an important part of this. There are many complex bioinformatics packages in the wild, but often they require a substantial amount of tinkering to work properly.\n",
    "\n",
    "You can (and should) use and modify the [code from the article](https://github.com/boscoh/pdbremix/blob/master/pdbremix/rmsd.py) to help parse PDB files and perform the alignments.  We are only interested in the core PDB parsing and RMSD calculations in this code.\n",
    "\n",
    "For each question below, you only need to consider the alpha carbon coordinates to represent the center of mass of the amino acid.  That is, use only one atom to represent each amino acid. \n",
    "\n",
    "Your program should take input as follows:\n",
    "```\n",
    "align.py -reference file1.pdb -mobile file2.pdb [-local window] -out outfile.pdb\n",
    "```\n",
    "\n",
    "Each solution should print the following to standard output:\n",
    "\n",
    "* RMSD calculation of the alignment\n",
    "* The rotation matrix that gives the best RMSD\n",
    "* A list of aligned residues in the following format (`A1->B1`,`A2->B2`, etc.)\n",
    "\n",
    "The `outfile.pdb` should be a new file (in PDB format) of the aligned structure.  Note that in an alignment, one structure is kept as reference and one is mobile.  The file should represent the mobile structure.  _You should apply the rotation matrix to the entire PDB, not just the alpha carbon atoms._"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the global RMSD betwen the follwoing myoglobin structures: 1mbn, 1np4, 3qm9, 4nos. Include the output from your program below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the best local RMSD alignment between the follwoing myoglobin structures (1mbn, 1np4, 3qm9, 4nos) by comparing segments of length 10. You should apply the rotation matrix to the entire protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the best structural alignment between residues in the heme binding pockets. Use both global and local approaches to find the best alignment.\n",
    "\n",
    "Use the [HEM binding pocket from 1mbn](https://googledrive.com/host/0B3XzcKIiWyccaUFjZ2RjTndyd2c//MPCS56420/2014-Autumn/2014-Autumn-Assignments/1mbn.HEM_A_155.pdb) as the query binding pocket.  Compare the query against each of the following binding pockets. \n",
    "\n",
    "* [HEM binding pocket from 1np4](https://googledrive.com/host/0B3XzcKIiWyccaUFjZ2RjTndyd2c//MPCS56420/2014-Autumn/2014-Autumn-Assignments/1np4.HEM_A_185.pdb)\n",
    "\n",
    "* [HEM binding pocket from 3qm9](https://googledrive.com/host/0B3XzcKIiWyccaUFjZ2RjTndyd2c//MPCS56420/2014-Autumn/2014-Autumn-Assignments/3qm9.HEM_A_201.pdb)\n",
    "\n",
    "* [HEM binding pocket from 4nos](https://googledrive.com/host/0B3XzcKIiWyccaUFjZ2RjTndyd2c//MPCS56420/2014-Autumn/2014-Autumn-Assignments/4nos.HEM_A_510.pdb)\n",
    "\n",
    "How does the similarity between the HEME binding pockets compare to the overall sequence and structural similarity? Report the sequence similarity in addition to the structural similarity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
