{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Speedup is a metric used to assess the relative performance improvement gained by executing one task versus another.  In our case, we are comparing serial execution (a single process) of searching a single sequence against a database versus a task parallelism version. \n",
    "\n",
    "Speedup is defined as \n",
    "$$ S \\equiv \\frac{T_{\\text{Old}}}{T_{\\text{New}}} $$ \n",
    "\n",
    "where $S$ is speedup, $T_{\\text{Old}}$ is the time taken to execute the script without improvement (serial), and $T_{\\text{New}}$ is the time taken to execute the script with the improvement (parallel).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "What is the speedup of the best performing `mpiBlast` runs compared to a serial Blast run on a *single node using a single thread*?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Update your code for local and global sequence alignment to compute all vs. all comparisions of a database.  The application will take adavantage of parallel computing modes in a high-performance computing environment.  Given a database of sequences, you application will compare every sequence against everyother sequence in the database (eg. all-vs-all).  The scores will be recorded and ouput as a single file and a matrix.\n",
    "\n",
    "Your code should take input from command line to specify the following parameters:\n",
    " - database file (in FASTA format)\n",
    " - gap penalty\n",
    " - scoring matrix (options should be PAM250 or BLOSUM62)\n",
    " - output file name\n",
    " - the number of tasks to submit\n",
    " \n",
    " For example:\n",
    " ```\n",
    " all-vs-all.py --database protein.db --task 8 --scoring_matrix PAM250 --gap_penalty -5 --output_file protein_score.txt \n",
    " ```\n",
    "\n",
    "The final output of your program should be 2 files.  The first should contain the description line for a seqeunce, the query and the alignment score.  For example:\n",
    "\n",
    "```\n",
    ">proteinA | >proteinB | Score\n",
    "```\n",
    "\n",
    "The second should contain the description line for a seqeunce and alignment scores in a matrix.  For example:\n",
    "\n",
    "```\n",
    "           >proteinA >proteinB >proteinC >proteinD\n",
    ">proteinA       #         #         #         #\n",
    ">proteinB       #         #         #         #\n",
    ">proteinC       #         #         #         #\n",
    ">proteinD       #         #         #         #\n",
    "```\n",
    "\n",
    "Please contain the description line in double quotes.  For exammple\n",
    "``` \n",
    ">XXX1223E4 | PROTEIN A | ECOLI\n",
    "```\n",
    "should be \n",
    "```\n",
    "\">XXX1223E4 | PROTEIN A | ECOLI\"\n",
    "```\n",
    "\n",
    "The applications should perform the following functions:\n",
    " * Split the database based on the number of tasks\n",
    " * Load balance the database so each bin is approximately equal\n",
    " * Perform the sequence alignments\n",
    " * Collate and sort the results\n",
    " * Write results to file\n",
    "\n",
    "The strategy you use to accomplish this is up to you.  You may use any approach discussed in class (e.g. mpi, openMP, embarassingly parallel, hybrid, etc.).  The application should be able to be launched from a single command. \n",
    "\n",
    "You can use the following [database](https://raw.githubusercontent.com/uchicago-bio/RCC-Utilities/master/database/pdb1000.fasta) to test your application.  It is a selection of proteins from the Protein Databank."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identify a gene you are interested in studying for your gene presentation. Provide a link to NCBI.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Protein Structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal of this question is to familiarize yourself with protein structures and to visualize 3D models of proteins using PyMOL.\n",
    "\n",
    "* Download and install PyMOL (https://www.pymol.org/) on your computer. You can download the installer or use Anaconda: `conda install -c schrodinger pymol`. You will not need a license to run it.\n",
    "* Consult the PyMOL wiki (https://pymolwiki.org) for answers to all your questions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing Proteins \n",
    "Download a protein from the PDB within Pymol interactive console.\n",
    "```\n",
    "PyMOL>fetch 1mbn\n",
    "```\n",
    "- What is this protein? \n",
    "- What organism does it come from? \n",
    "- Provide the URL to the protein from the PDB website?\n",
    "- Take a screenshot and include it below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you load a structure, the default visualization in PyMOL is a `cartoon` represenation. You can highlight the secondary strucure elements by coloring by secondary structure elements by selecting `C -> by ss -> Helix Sheet Loop`.\n",
    "- How many alpha-helices does the protein have? \n",
    "- How many beta sheets does it have?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While this visualization makes for great images for magazine covers, it does not provide the detail we may need to analyze it. In particular, the `licorice sticks` view allows us to see an all-atom reprentation that will help us view the side chains. Change the view in the GUI by selecting `H -> everything` and then `S -> sticks`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer, you can also control the depiction from the command line.\n",
    "```\n",
    "PyMOL> hide everything, 1mbn\n",
    "PyMOL> show sticks, 1mbn\n",
    "```\n",
    "Take a screenshot and include it below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparing Proteins\n",
    "\n",
    "Download a different protein from the PDB using the following commands.\n",
    "```\n",
    "PyMOL> fetch 1ase\n",
    "PyMOL> hide everything \n",
    "PyMOL> show cartoon\n",
    "```\n",
    "\n",
    "- What is this protein?\n",
    "- How many alpha-helices does the protein have?\n",
    "- How many beta sheets does it have?\n",
    "- What are the ligands (small molecules not part of the protein) that is bound to it?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This protein was part of a functional study that included mutating residues. Load in another version of the protein with a mutation in residue 226.\n",
    "\n",
    "```\n",
    "PyMOL> fetch 1asf\n",
    "PyMOL> align 1ase, 1asf\n",
    "PyMOL> select mutant, resi 226\n",
    "PyMol> show sticks, mutant\n",
    "```\n",
    "\n",
    "- What is the mutation that was made in the protein? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complete PyMOL Tutorial from PyRosetta: https://graylab.jhu.edu/pyrosetta/downloads/documentation/pyrosetta4_online_format/PyRosetta4_Workshop1_PyMOL.pdf PyRosetta is a powerful molecular modeling toolkit that we will be using later on in the course."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
